Libtool Conan 2.0 compatibility

System-Arch opened the pull request:
Specify library name and version:  **libtool/2.4.7**

I acknowledge that there is a pre-existing effort to modernize libtool in #14466 in support of #12888, however, I had a more immediate need for Conan 2.0 support and was concerned that the existing efforts were going to need additional refinements to address issues like https://github.com/conan-io/conan/issues/12785 and https://github.com/conan-io/conan/issues/12812, so I took my learnings from those conversations along with the technique applied by @jcar87 in #15245 and tried applying it here. It also served as a further learning experience as getting the test_package to work was one of the more difficult porting tasks I have taken since starting my Conan activities. Thus, at minimum, I hope it can serve as a resource for documenting some of the challenges that one can encounter with complex library combinations.


---

- [X] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [X] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

prince-chrismc commented:
> learning experience as getting the test_package to work was one of the more difficult porting tasks I have taken since starting my Conan activities. 

We absolutely appreciate this ❤️

System-Arch commented:
I acknowledge adding conditional code based on Conan version in the test_package recipe in order to get it to work with both Conan 2.0 and Conan 1.x despite it being frowned upon. My rationale (as also noted above) is to document what it takes to make these tests work. The conditional changes are related to the differences in how Conan 1.x and Conan 2.0 handle self.package_folder when running the test command. Conan 1.x leaves package_folder undefined whereas Conan 2.0 sets it to test_package/test_ouput. The benefit of the Conan 2.0 approach is that it permits the use of cmake.install(). On the other hand, the behavior of cmake.install() is inconsistent with autotools.install() in that autotools.install() implicitly permits the specification of the package directory via an override of DESTDIR, whereas cmake.install() has no such provision. This test suite would have been cleaner and easier to comprehend if cmake_install() permitted the pkg_folder argument to --premix to be overridden in cmake.py versus how it is currently pegged to package_folder:

>         pkg_folder = '"{}"'.format(self._conanfile.package_folder.replace("\\", "/"))
>         build_folder = '"{}"'.format(self._conanfile.build_folder)
>         arg_list = ["--install", build_folder, build_config, "--prefix", pkg_folder]

If pkg_folder was permitted as an optional named argument, I posit that the Conan 1.x behavior of providing no default value for package_folder in the text context might make more sense, thereby forcing one to define it explicitly if using cmake_install in a test package the same way that one must override DESTDIR in this context to obtain meaningful results with autotools. Pinging @memsharded in case he has additional thoughts or insights to provide here.

jcar87 commented:
Hi @System-Arch  - thanks so much for this PR again. This one is on my list for tomorrow to review :)

System-Arch commented:
> Hi @System-Arch - thanks so much for this PR again. This one is on my list for tomorrow to review :)

Hi @jcar87,
Thanks for your message. After several more days of work, I finally got both test suites to pass with Conan 1.x, which I must say has been incredibly difficult in comparison to Conan 2.,0. Note also that the Conan 2.0 CI failures are due to missing dependencies in the 2.0 repo. I'm not sure why things like m4 are missing since the recipes for all of the dependencies are working with Conan 2.0 at our site (admittedly, we are leveraging all of my pending PRs). Thanks again for your help on getting us all migrated to Conan 2.0

Pro commented:
@System-Arch I tried you recipe, and unfortunately was hitting #9002

I.e., the output of the build fails with:

```
   Creating library libltdl\.libs\libltdl.dll.lib and object libltdl\.libs\libltdl.dll.exp
libltdl_la-ltdl.obj : error LNK2019: unresolved external symbol lt_ltdl_LTX_preloaded_symbols referenced in function lt_dlinit
libltdl\.libs\libltdl-7.dll : fatal error LNK1120: 1 unresolved externals
make[2]: *** [Makefile:1172: libltdl/libltdl.la] Error 2
make[2]: Leaving directory '/c/.conan/5b40cf/1/build-release'
make[1]: *** [Makefile:1648: all-recursive] Error 1
make[1]: Leaving directory '/c/.conan/5b40cf/1/build-release'
make: *** [Makefile:1025: all] Error 2
libtool/2.4.7@sp/test:
libtool/2.4.7@sp/test: ERROR: Package '5a61a86bb3e07ce4262c80e1510f9c05e9b6d48b' build failed
libtool/2.4.7@sp/test: WARN: Build folder C:\.conan\5b40cf\1\build-release
ERROR: libtool/2.4.7@sp/test: Error in build() method, line 94
        autotools.make()
        ConanException: Error 2 while executing make -j8
```

If I remove the `--enable-shared` option from the conanfile.py build on VS 2022 succeeds (test_package still fails though, also needs adaption).

```patch
diff --git a/recipes/libtool/all/conanfile.py b/recipes/libtool/all/conanfile.py
index 50bba3fee..7b67f2876 100644
--- a/recipes/libtool/all/conanfile.py
+++ b/recipes/libtool/all/conanfile.py
@@ -69,7 +69,6 @@ class LibtoolConan(ConanFile):
         if (self.settings.get_safe("compiler") == "Visual Studio" and self.settings.get_safe("compiler.version") >= "12") or \
            (self.settings.get_safe("compiler") == "msvc" and self.settings.get_safe("compiler.version") >= "180"):
             tc.extra_cflags.append("-FS")
-        tc.configure_args.append("--enable-shared")
         tc.configure_args.append("--enable-static")
         tc.configure_args.append("--enable-ltdl-install")
         env = tc.environment()

```


Is the `--enable-shared` option needed?

Would you know any reason why we get the unresolved external symbol?

Pro commented:
Update to my previous comment to provide a proper fix for #9002:

The correct solution can be found here:
https://savannah.gnu.org/patch/?9313#comment1

I.e., the following two vars need to be set: F77=no and FC=no

With the following patch, the whole recipe, including test packages work on VS2022:

```patch
diff --git a/recipes/libtool/all/conanfile.py b/recipes/libtool/all/conanfile.py
index 50bba3fee..833c1bbb3 100644
--- a/recipes/libtool/all/conanfile.py
+++ b/recipes/libtool/all/conanfile.py
@@ -78,6 +78,8 @@ class LibtoolConan(ConanFile):
             env.append("CXX", f'{unix_path(self, self.conf.get("user.automake:compile-wrapper"))} cl -nologo')
             env.append("AR", f'{unix_path(self, self.conf.get("user.automake:lib-wrapper"))} lib')
             env.append("LD", "link")
+            env.define("F77", "no")
+            env.define("FC", "no")
         tc.generate(env)

     def _patch_sources(self):
```

Note: I tested with `env.append("F77", "no")` but there I still had the original issue. So, only `env.define` seems to work

kevinchannon commented on recipes/libtool/all/conanfile.py:
```diff
@@ -26,106 +31,80 @@ class LibtoolConan(ConanFile):
+            env.append("CC", f'{unix_path(self, self.conf.get("user.automake:compile-wrapper"))} cl -nologo')
+            env.append("CXX", f'{unix_path(self, self.conf.get("user.automake:compile-wrapper"))} cl -nologo')
+            env.append("AR", f'{unix_path(self, self.conf.get("user.automake:lib-wrapper"))} lib')
+            env.append("LD", "link")
```
If this is a fix for the specific issue that the libtool package fails to build on Windows via MSVC&mdash;which I'm currently hitting :)&mdash;would it be possible to create a separate PR for it?  It's just that this PR is currently failing the non-essential Conan 2.0 build checks. This type of failure looks like it would normally be non-blocking, but this PR is specifically about upgrading the package to work with Conan 2.0, so it seems like it could block the PR.

If this small change can get the package to work for Windows MSVC users, then it seems like a quick win to just merge it in isolation, without having to wait for the huge pile of changes in the current PR to be fixed for Conan 2.0 and merged. Right?

(Full disclosure: this is my first look at any PR on conan-io, so there could be a thousand good reasons why the above suggestion is laughably nonsensical. Apologies if that's the case!)

jcar87 replied:
Hi @kevinchannon  - absolutely! 
if I'm not mistaken, this is a specific fix to build on Windows with Visual Studio 2022 - the recipe as it is should still builds with earlier versions (without the need for this change). If that isn't the case, please let me know!

I'm aiming to have this PR reviewed and merged sometime tomorrow, including this fix, but otherwise feel free to open a PR just for the changes to this issue, referencing the upstream thread https://savannah.gnu.org/patch/?9313#comment1, so that we know that the workaround can be removed in a future version if they decide to accept the patch. I think these two lines disable fortran compiler detection

jcar87 commented on recipes/libtool/all/patches/2.4.6-0001-libtool-relocatable.patch:
```diff
@@ -3,20 +3,27 @@
+@@ -1901,11 +1901,18 @@
    pkgmacro_files="@pkgmacro_files@"
 
++  if [ "$(uname -s)" = "Darwin" ]; then
```
I think this may be due to bash/zsh things, rather than MacOS specific - I have some fixes for this, will post shortly

System-Arch replied:
Hi @jcar87,
I believe the lack of support for the UNIX command realpath(1) is a well-known limitation of BSD UNIX, which MacOS is derived from. Fortunately, it appears that Apple has added it with the recent release of Ventura. I have updated the logic here to use the OS-provided version when available. Thanks

jcar87 replied:
These are the contents of the patch I'm working with at the moment:

```diff
diff --git a/Makefile.in b/Makefile.in
index 29db7be..3046cdf 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -2423,10 +2423,10 @@ $(notes_txt): $(notes_texi)
 # files created in the build tree, so instead we regenerate the
 # manual pages if the sources for the build-tree files we want to
 # run have changed.
-$(libtool_1): $(ltmain_sh)
-	$(AM_V_GEN)$(update_mans) --help-option=--help-all libtool
-$(libtoolize_1): $(libtoolize_in)
-	$(AM_V_GEN)$(update_mans) libtoolize
+#$(libtool_1): $(ltmain_sh)
+#	$(AM_V_GEN)$(update_mans) --help-option=--help-all libtool
+#$(libtoolize_1): $(libtoolize_in)
+#	$(AM_V_GEN)$(update_mans) libtoolize
 
 install-data-local: $(lt_Makefile_in) install-scripts-local
 	@$(NORMAL_INSTALL)
diff --git a/libtoolize.in b/libtoolize.in
index 0c40fed..e82e467 100644
--- a/libtoolize.in
+++ b/libtoolize.in
@@ -1903,11 +1903,12 @@ func_require_seen_libtool ()
   pkgmacro_files="@pkgmacro_files@"
 
   # Locations for important files:
-  prefix="@prefix@"
-  datadir="@datadir@"
-  pkgauxdir="@pkgauxdir@"
-  pkgltdldir="@pkgdatadir@"
-  aclocaldir="@aclocaldir@"
+  scriptdir=$(CDPATH= cd -- "$(dirname -- "$0")" && pwd)
+  prefix="$scriptdir/.."
+  datadir="$scriptdir/../res"
+  pkgauxdir="$scriptdir/../res/libtool/build-aux"
+  pkgltdldir="$scriptdir/../res/libtool"
+  aclocaldir="$scriptdir/../res/aclocal"
 
   # Allow the user to override the master libtoolize repository:
   if test -n "$_lt_pkgdatadir"; then
```

this seems to work on the platforms I've tested it without having to use a different case on macOS - although I need to test it on an older version of macOS as well.

jcar87 commented on recipes/libtool/all/conanfile.py:
```diff
@@ -220,22 +202,36 @@ def package_info(self):
-        for key, value in self._libtool_relocatable_env.items():
-            self.output.info("Setting {} environment variable to {}".format(key, value))
+        libtool_relocatable_env = {
+            "LIBTOOL_PREFIX": self.package_folder,
```
If the package is relocatable as per the patch, these are not needed anymore - these variables, confusingly, were a construct of the recipe and not part of the upstream code.

System-Arch replied:
Hi @jcar87,
Good point.  I have removed them per your suggestion,. Thanks

jcar87 commented on recipes/libtool/all/conanfile.py:
```diff
@@ -220,22 +202,36 @@ def package_info(self):
-        self.env_info.ACLOCAL_PATH.append(libtool_aclocal)
-        self.output.info("Appending AUTOMAKE_CONAN_INCLUDES environment variable: {}".format(libtool_aclocal))
-        self.env_info.AUTOMAKE_CONAN_INCLUDES.append(libtool_aclocal)
+        aclocal_path = "/" + aclocal_path.replace("\\", "/").replace(":", "") # Can't use unix_path with Conan 2.0
```
we have added a helper function to cover cases were `unix_path` was erroneously used in `package_info`, to aid with the migration -

System-Arch replied:
Hi @jcar87,
Okay--I replaced the above line with a call to unix_path_package_info_legacy. Thanks

jcar87 reviewed (changes requested):
this is really good!! and works well :D 
Some minor changes - I'm happy to push those myself or branch off from this one :D we're almost there

jcar87 commented:
I'm currently working on this recipe, will open a new PR based off the changes proposed here :D Just minor things really - currently testing across all platforms before raising a new one

System-Arch commented:
> I'm currently working on this recipe, will open a new PR based off the changes proposed here :D Just minor things really - currently testing across all platforms before raising a new one

Hi @jcar87,
I believe I have implemented all of the changes you have suggested so it's probably not necessary to open a new PR. Unfortunately the CCI build pipeline is down so it's difficult to validate them. Thanks

jcar87 commented:
Hi @System-Arch  - thank you so much for implementing my suggestions. There were a few more things (apologise I wasn't able to write a more thorough review), so I've gone ahead and opened a new PR here, based upon your changes: https://github.com/conan-io/conan-center-index/pull/16214

A summary of the changes on top of your branch:

* In the patch file, use `scriptdir=$(CDPATH= cd -- "$(dirname -- "$0")" && pwd)` to determine the location of the scripts, and derive the rest of the directories relative to this. Have tested this on Windows, Linux and Mac - please let me know if this covers your use case on macOS that required a different alternative.
* set the package type to "library" - while libtool is mainly use as build-time scripts ("application") - it is sometimes used as a library (for libltdl) and the traits for "library" are already a superset of application, so this should cover all use cases
* have retained the logic (from the recipe before our changes) so conditionally use `tool_requires` on the presence of `build_settings` (i.e., using two profiles) - indeed it would be preferable to do so unconditionally, however have preferred to not make that change if the recipe already had this conditionally, for Conan 1.x users still using a single profile
* have retained the `datarootdir` logic, this one works if we pass `"--datarootdir=${prefix}/res"` to the autotools configure args - I believe this means we don't have to remove/rename things manually in the `package()` method 
* modernised the conditional check to pass the `-FS` flag when using msvc
* cleaned up the logic for defining the necessary environment variables (indeed if the package is self-relocatable, they are no longer needed, except for the aclocal ones) 


There were a few things I didn't fully understand
* I've removed `if not is_msvc(self): # Don't delete Windows import lib` - was a file erroneously being deleted?
*  I've removed `self.cpp_info.set_property("pkg_config_name", "libtool")` unless there's a specific usecase we're not covering? would love to know more
* I've removed ` self.cpp_info.resdirs = ["res/aclocal", "res/libtool"]` out of an abudance of caution - they are already exposed to autotools in the aclocal environment variables (and otherwise referenced internally in the bash scripts), so I don't expect users to have to "publicly" query where these are from outside of the recipe, although please correct me if there's a use-case that I'm missing


the test package was very complicated already before this PR, so thanks so much for your efforts to port it. I've made a few changes, hopefully to simplify the logic:
* avoid code duplication when instantiating the two autotoolstoolchains (by saving the msvc-related variables in a dictionary)
* define the runtime `*_LIBRARY_PATH` variables in the `generate()` method so that the script is loaded by the conanrun script, and the logic in `self.run` can be simplified
* in `test_ltdl`, use `self.cpp.build.bindirs[0]` which should work in all cases without having to use logic depending on conan version
* I believe calling `cmake.install()` for the first cmake project (`ltdl`) is not necessary
* to avoid the cmaketoolchain/cmake reusing the same build directory for two completely different cmake projects (and indeed avoid the need to clean the cmake cache), I've moved the static library to the same cmake project, and I'm calling `cmake --install` directly when testing static-in-shared, to install only that static library
* to simplify logic of deriving filename and extensions, I have changed the "loadable" library to `MODULE` in cmake (which is indeed intended for libraries loaded at runtime via `dlopen`). This makes the extension `.so` on macOS, which allows us to simplify the logic altogether.


Hopefully this is clear, let me know if you have any questions!! And once again, thanks so much for porting this recipe :D

System-Arch commented:
> Hi @System-Arch - thank you so much for implementing my suggestions. There were a few more things (apologise I wasn't able to write a more thorough review), so I've gone ahead and opened a new PR here, based upon your changes: #16214

Hi @jcar87,

Thanks for your detailed write-up. Please see the comments below:

> There were a few things I didn't fully understand
> 
> * I've removed `if not is_msvc(self): # Don't delete Windows import lib` - was a file erroneously being deleted?

Yes, the Windows import lib was being deleted when building ltdl as a DLL.

> * I've removed `self.cpp_info.set_property("pkg_config_name", "libtool")` unless there's a specific usecase we're not covering? would love to know more

No, I thought it was a standard a pattern so I just included it for consistency. If it's not needed for anything, it's fine to leave it out.

> * I've removed ` self.cpp_info.resdirs = ["res/aclocal", "res/libtool"]` out of an abudance of caution - they are already exposed to autotools in the aclocal environment variables (and otherwise referenced internally in the bash scripts), so I don't expect users to have to "publicly" query where these are from outside of the recipe, although please correct me if there's a use-case that I'm missing

I'm not sure. It took me days of trial and error to get these tests to pass with Conan 1.x (vastly more effort than Conan 2.0). It's possible that these are not needed in light of other changes. 

> * avoid code duplication when instantiating the two autotoolstoolchains (by saving the msvc-related variables in a dictionary)

Yes, I know the multiple autotoolchains were redundant. Part of my motivation for getting this package ported was to create examples of functional patterns. In this instance, it was illustrating how to use namespaces. At present most of the Conan 2.0 documentation is pretty thin so I typically find myself reading the Python source code to deduce how things work.

> Hopefully this is clear, let me know if you have any questions!! And once again, thanks so much for porting this recipe :D

All of the other changes sound reasonable. Unfortunately, I was asked to remove the support for autogen in #14687 as it was my system-level proof that my collection of autotools changes (autoconf, automake, libtool, autoconf-archive) worked with all supported OS's with both Conan 1.x and Conan 2.0. Maybe you can help persuade the powers-that-be of the value of that work. Thanks,

jcar87 commented:
Thanks for taking the time to read through my notes!

> Yes, the Windows import lib was being deleted when building ltdl as a DLL.

wow, thanks for figuring this out! I'll restore this :D 

> Yes, I know the multiple autotoolchains were redundant. Part of my motivation for getting this package ported was to create examples of functional patterns. In this instance, it was illustrating how to use namespaces. At present most of the Conan 2.0 documentation is pretty thin so I typically find myself reading the Python source code to deduce how things work.

Oh, the two toolchains with namespaces were not reduntant at all - in fact, it was the right thing to do, as in particular the `sis` project required additional configure arguments, and thus, its own toolchain :D 
The bit I was talking about was the duplication of the msvc logic - I've made an attempt to simplify that a tiny bit.

samuel-emrys commented:
>>I've removed self.cpp_info.set_property("pkg_config_name", "libtool") unless there's a specific usecase we're not covering? would love to know more

> No, I thought it was a standard a pattern so I just included it for consistency. If it's not needed for anything, it's fine to leave it out.

I've been advised to keep these statements in in the past for reviewing purposes. See https://github.com/conan-io/conan-center-index/pull/15109#discussion_r1090333625

samuel-emrys commented:
Obviously difficult in light of the CI being down for maintenance, but might I suggest that we expedite the approval of this recipe since it blocks the migration of a number of other well used packages (i.e., libcurl, date, binutils, libffi, libiconv, libuuid, more), and make any improvements in subsequent PRs?

jcar87 commented:
> Obviously difficult in light of the CI being down for maintenance, but given that 2.0 was just released, might I suggest that we expedite the approval of this recipe since it blocks the migration of a number of other well used packages (i.e., libcurl, date, binutils, libffi, libiconv, libuuid, more), and make any improvements in subsequent PRs?

Hi @samuel-emrys - indeed, this is the highest priority for today and I have a branch based off this one that is already undergoing CI. I will keep you posted.

System-Arch commented:
Superseded by #16248
